lib/src/voip/utils/types.dart

```diff
@@ -0,0 +1,180 @@
+
+// Call state
+enum CallState {
+  /// The call is inilalized but not yet started
+  kFledgling,
+
+  /// The first time an invite is sent, the local has createdOffer
+  kInviteSent,
+
+  /// getUserMedia or getDisplayMedia has been called,
+  /// but MediaStream has not yet been returned
+  kWaitLocalMedia,
+
+  /// The local has createdOffer
+  kCreateOffer,
+
+  /// Received a remote offer message and created a local Answer
+  kCreateAnswer,
+
+  /// Answer sdp is set, but ice is not connected
+  kConnecting,
+
+  /// WebRTC media stream is connected
+  kConnected,
+
+  /// The call was received, but no processing has been done yet.
+  kRinging,
+
+  /// End of call
+  kEnded,
+}
+
+class CallErrorCode {
+  /// The user chose to end the call
+  static String UserHangup = 'user_hangup';
+
+  /// An error code when the local client failed to create an offer.
+  static String LocalOfferFailed = 'local_offer_failed';
+
+  /// An error code when there is no local mic/camera to use. This may be because
+  /// the hardware isn't plugged in, or the user has explicitly denied access.
+  static String NoUserMedia = 'no_user_media';
+
+  /// Error code used when a call event failed to send
+  /// because unknown devices were present in the room
+  static String UnknownDevices = 'unknown_devices';
+
+  /// Error code used when we fail to send the invite
+  /// for some reason other than there being unknown devices
+  static String SendInvite = 'send_invite';
+
+  /// An answer could not be created
+
+  static String CreateAnswer = 'create_answer';
+
+  /// Error code used when we fail to send the answer
+  /// for some reason other than there being unknown devices
+
+  static String SendAnswer = 'send_answer';
+
+  /// The session description from the other side could not be set
+  static String SetRemoteDescription = 'set_remote_description';
+
+  /// The session description from this side could not be set
+  static String SetLocalDescription = 'set_local_description';
+
+  /// A different device answered the call
+  static String AnsweredElsewhere = 'answered_elsewhere';
+
+  /// No media connection could be established to the other party
+  static String IceFailed = 'ice_failed';
+
+  /// The invite timed out whilst waiting for an answer
+  static String InviteTimeout = 'invite_timeout';
+
+  /// The call was replaced by another call
+  static String Replaced = 'replaced';
+
+  /// Signalling for the call could not be sent (other than the initial invite)
+  static String SignallingFailed = 'signalling_timeout';
+
+  /// The remote party is busy
+  static String UserBusy = 'user_busy';
+
+  /// We transferred the call off to somewhere else
+  static String Transfered = 'transferred';
+}
+
+class CallError extends Error {
+  final String code;
+  final String msg;
+  final dynamic err;
+  CallError(this.code, this.msg, this.err);
+
+  @override
+  String toString() {
+    return '[$code] $msg, err: ${err.toString()}';
+  }
+}
+
+enum CallEvent {
```

Comment:
This is quite a confusing name.

Reply:
You have several definitions, for what a "CallEvent" is. We should probably use different names for Events for calls and events for call state changes. Especially since we already have a "CallState" that maps very closely to this. How about naming this a "CallChange" instead? Or "CallStateChange"? I know that "CallEvent" isn't wrong, but it is very confusing when you already have Matrix events for calls.

Reply:
makes sense, td likes CallStateChange changed